presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java

```diff
@@ -263,57 +268,87 @@ public void addSources(List<TaskSource> sources)
     {
         Map<PlanNodeId, TaskSource> updatedUnpartitionedSources = new HashMap<>();
 
-        // don't update maxAcknowledgedSplit until the end because task sources may not
-        // be in sorted order and if we updated early we could skip splits
-        long newMaxAcknowledgedSplit = maxAcknowledgedSplit;
+        // first remove any split that was already acknowledged
+        sources = sources.stream()
+                .map(source -> new TaskSource(
+                        source.getPlanNodeId(),
+                        source.getSplits().stream()
+                                .filter(scheduledSplit -> scheduledSplit.getSequenceId() > maxAcknowledgedSplit)
+                                .collect(Collectors.toSet()),
+                        source.isNoMoreSplits()))
+                .collect(toList());
 
         for (TaskSource source : sources) {
-            PlanNodeId sourceId = source.getPlanNodeId();
-            if (sourceId.equals(partitionedSourceId)) {
-                // partitioned split
-                ImmutableList.Builder<DriverSplitRunner> runners = ImmutableList.builder();
-                for (ScheduledSplit scheduledSplit : source.getSplits()) {
-                    // only add a split if we have not already scheduled it
-                    if (scheduledSplit.getSequenceId() > maxAcknowledgedSplit) {
-                        // create a new driver for the split
-                        runners.add(partitionedDriverFactory.createDriverRunner(scheduledSplit, true));
-                        newMaxAcknowledgedSplit = max(scheduledSplit.getSequenceId(), newMaxAcknowledgedSplit);
-                    }
-                }
-
-                enqueueDrivers(false, runners.build());
-                if (source.isNoMoreSplits()) {
-                    partitionedDriverFactory.setNoMoreSplits();
-                }
+            if (partitionedDriverFactories.containsKey(source.getPlanNodeId())) {
+                schedulePartitionedSource(source);
             }
             else {
-                // unpartitioned split
+                scheduleUnpartitionedSource(source, updatedUnpartitionedSources);
+            }
+        }
 
-                // update newMaxAcknowledgedSplit
-                for (ScheduledSplit scheduledSplit : source.getSplits()) {
-                    newMaxAcknowledgedSplit = max(scheduledSplit.getSequenceId(), newMaxAcknowledgedSplit);
-                }
+        // update maxAcknowledgedSplit
+        maxAcknowledgedSplit = sources.stream()
+                .flatMap(source -> source.getSplits().stream())
+                .mapToLong(ScheduledSplit::getSequenceId)
+                .max()
+                .orElse(maxAcknowledgedSplit);
+        return updatedUnpartitionedSources;
```

Comment:
What is unpartitioned source?

Unrelated question: What's going on in the caller of this code? weak reference, etc.